REVIEW [NEXUS-5672] Create an ExtJS3 UI plugin

nabcos opened the pull request:
https://issues.sonatype.org/browse/NEXUS-5672

This pull request collects most scattered UI resources into one plugin. Only UiContributor/IndexHtmlCustomizer remains in nexus-core and is changed to honor the `?debug` URL parameter. Plugin javascript will then be loaded in the single (dev) files instead of the minified version. By using XmlPluginRepository + symlinking resources/static/js to target/classes/static/js you can edit the js in place and only need to reload the Nexus UI, no need to restart Nexus for UI dev anymore.

This pull changes setup for yuicompressor plugin, so every plugin contributing UI will need to move their js/css files to src/main/resources/static/(js|css) and needs a dependency on the new UI plugin to pick up UiContributor/IndexHtmlCustomizer.

Next step in terms of UI maintenance/improvements would be to add something like https://github.com/pguedes/maven-jsduck to produce js documentation. That plugin is working, but it seems like that particular plugin is not published into a public repository ATM, so I refrained from adding it to the build for now (there's not much documentation in the current sources anyway).

nabcos committed:
Include js sources for ?debug mode

nabcos committed:
Use non-deprecated property

nabcos committed:
Move redirector resource for /->/index.html

nabcos committed:
Move Sonatype-content.css to restlet plugin for now (content/error page)

nabcos committed:
Redirect /images to /static/images

Exported location for resoures in plugins is /static. References to
/images are scattered around all plugins. Redirect for compatibility.

nabcos committed:
Remove UI in nexus-webapp

nabcos committed:
Add UI resources

nabcos committed:
Add js-lib

nabcos committed:
Add js-nx

nabcos committed:
Add UI app

nabcos committed:
Enable debug mode for UiContributorBuilder

nabcos committed:
Only redirect images if they are not found

nabcos committed:
Add config to copy js files to /static/js to plugin-parent

nabcos committed:
Remove unused dependency

nabcos committed:
Add ui plugin to dist

jdillon commented:
Curious why UiContributor/IndexHtmlCustomizer left in core?

nabcos committed:
Move UiContrib and related classes from core to plugin

nabcos committed:
move js/css to resources/static

kellyrob99 commented on plugins/nexus-ui-extjs3-plugin/src/main/css/Sonatype.css:
```diff
@@ -197,7 +197,7 @@ button.st-icon-refresh {
 
 .form-label-helpmark-combo {
     /* large margin needed to place helpmark next to combobox - the box is position:absolute, and helpmark was covered by dropdown arrow */
-    margin-left: 19px;
```
Appear to have overwritten a change here IIRC

nabcos replied:
I will probably rebase and redo the "move js/css" commits before merge, so that the latest changes from master are picked up here.

(I don't think git can handle a merge from master directly)

kellyrob99 commented on plugins/nexus-ui-extjs3-plugin/src/main/java/org/sonatype/nexus/plugins/ui/RedirectingNexusApplicationCustomizer.java:
```diff
@@ -0,0 +1,71 @@
+import org.sonatype.plexus.rest.RetargetableRestlet;
+
+/**
+ * Installs a filter before the NexusApplication root that will redirect /images to /static/images.
```
So this is to handle plugins that still load from /images path?

nabcos replied:
Yes, for compatibility reasons.

kellyrob99 commented:
I could use a little more context as to why this change is beneficial please, as it seems a bit counter-intuitive to be moving all of our js/css out of the actual webapp module?

nabcos committed:
Do not cache UiContributionBuilder

If it's used more than once, first hit with or without "?debug" will set
the state for the plugin

nabcos commented:
@kellyrob99 
- unify build (behaves just like any other plugin)
- easier UI development (just drop the plugin into any dist/use XmlPluginRepository instead of rebuild dist/unzip war contents/copy files)
- separate from unrelated plugins (UI needed parts in core, webapp, restlet plugin)
- make way for alternative implementation

kellyrob99 commented:
Thanks for the explanation Ben.
+1

jdillon commented:
+1 this seems to be steps in the right direction for current UI framework to support ?debug better.

nabcos committed:
Create UI plugin

?debug mode should work across plugins now, but the plugins are not
packaging the non-minified sources yet

nabcos committed:
Fix image paths

nabcos committed:
Fix redirect for images

nabcos committed:
Keep backward compatibility with plugins not using UiConributor

nabcos committed:
Fix license plugin config for static/resources/js

nabcos committed:
Add DM entry for ui plugin jar (idea profile)

nabcos committed:
Merge remote-tracking branch 'upstream/master' into extract-ui-plugin

* Small change to index.vm
